crates/red_knot/src/main.rs

```diff
@@ -133,3 +254,87 @@ fn main() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+enum Message {
+    AnalysisCompleted(Vec<String>),
+    AnalysisCancelled,
+    Exit,
+    FileChanges(Vec<FileChange>),
+}
+
+#[derive(Default, Debug)]
+struct AggregatedChanges {
+    changes: FxHashMap<FileId, FileChangeKind>,
+}
+
+impl AggregatedChanges {
+    fn add(&mut self, change: FileChange) {
+        match self.changes.entry(change.file_id()) {
+            Entry::Occupied(mut entry) => {
+                let merged = entry.get_mut();
+
+                match (merged, change.kind()) {
+                    (FileChangeKind::Created, FileChangeKind::Deleted) => {
+                        // Creation after deletion means that ruff newer saw the file.
+                        entry.remove();
+                    }
+                    (FileChangeKind::Created, FileChangeKind::Modified) => {
+                        // No-op, for ruff, modifying a file that it doesn't yet know that it exists is still considered a creation.
+                    }
+
+                    (FileChangeKind::Modified, FileChangeKind::Created) => {
+                        // Uhh, that should probably not happen. Continue considering it a modification.
```

Comment:
maybe better to assert here if we really think it shouldn't happen?

Reply:
Yeah, not sure. I don't think it's something I want to drip Ruff over. I think we probably want a `Change` that means `anything` happened to the file so that we can take the most defensive invalidation for that file.